### PR TITLE
[metrics] Update SSIM

### PR DESCRIPTION
**Pull request by ydcjeff**
## What does this PR do?

Update SSIM and tests to compare with lower absolute tolerance

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In in short, see following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by teddykoker** [comment]
Thanks for working on this!

**Comment by ydcjeff** [comment]
@teddykoker @SkafteNicki mind review a bit so far?
It has now squashed to 7e-5 tolerance

**Comment by SkafteNicki** [comment]
@ydcjeff the only clear difference I can find now between the two implementations seems to be some cropping done after the convolutions:
https://github.com/scikit-image/scikit-image/blob/ed889ba3deb17cd94d1f0f9b730f6564585936ee/skimage/metrics/_structural_similarity.py#L210-L214
Do you know if we should also do this?

**Comment by ydcjeff** [comment]
> @ydcjeff the only clear difference I can find now between the two implementations seems to be some cropping done after the convolutions:
> https://github.com/scikit-image/scikit-image/blob/ed889ba3deb17cd94d1f0f9b730f6564585936ee/skimage/metrics/_structural_similarity.py#L210-L214
> Do you know if we should also do this?

I think we should

File: tests/metrics/regression/test_ssim.py [hunk 1]
```diff
@@ -50,7 +51,7 @@ def _sk_metric(preds, target, data_range, multichannel):
     [(i.preds, i.target, i.multichannel) for i in _inputs],
 )
 class TestSSIM(MetricTester):
-    atol = 1e-3  # TODO: ideally tests should pass with lower tolerance
+    atol = 6e-5  # TODO: ideally tests should pass with lower tolerance
```
Comment (SkafteNicki) [inline tests/metrics/regression/test_ssim.py:54]: How tight is this bound? Right now we are seeing SSIM only failing sometime, because of randomness in input. I was wondering if we should set `atol=1e-5` just to be sure.

Comment (ydcjeff) [reply tests/metrics/regression/test_ssim.py:54]: I tested `1e-5` locally and it didn't pass, so increase to 6e-5

Comment (SkafteNicki) [reply tests/metrics/regression/test_ssim.py:54]: Sorry I was confusing myself. Just tried running the test many times (with different seeds), seems to be stable now. Nicely done.

**Review by SkafteNicki — APPROVED** [review]

File: tests/metrics/regression/test_ssim.py (hunk 1 above)
Comment (tchaton) [inline tests/metrics/regression/test_ssim.py:54]:
```suggestion
    atol = 6e-5
```

**Review by tchaton — APPROVED** [review]
LGTM ! I had a doubt with F.pad, but it is correct usage.

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/metrics/functional/ssim.py [hunk 2]
```diff
@@ -19,17 +19,17 @@
...
-def _gaussian_kernel(channel, kernel_size, sigma, device):
-    def _gaussian(kernel_size, sigma, device):
-        gauss = torch.arange(
-            start=(1 - kernel_size) / 2, end=(1 + kernel_size) / 2, step=1, dtype=torch.float32, device=device
-        )
-        gauss = torch.exp(-gauss.pow(2) / (2 * pow(sigma, 2)))
-        return (gauss / gauss.sum()).unsqueeze(dim=0)  # (1, kernel_size)
+def _gaussian(kernel_size: int, sigma: int, dtype: torch.dtype, device: torch.device):
+    dist = torch.arange(start=(1 - kernel_size) / 2, end=(1 + kernel_size) / 2, step=1, dtype=dtype, device=device)
+    gauss = torch.exp(-pow(dist / sigma, 2) / 2)
```
Comment (Borda) [inline pytorch_lightning/metrics/functional/ssim.py:24]:
```suggestion
    gauss = torch.exp(-torch.pow(dist / sigma, 2) / 2)
```

**Review by Borda — COMMENTED** [review]
lgtm, but I am not a metrics guy 🐰

File: tests/metrics/regression/test_ssim.py [hunk 3]
```diff
@@ -50,7 +51,7 @@ def _sk_metric(preds, target, data_range, multichannel):
     [(i.preds, i.target, i.multichannel) for i in _inputs],
 )
 class TestSSIM(MetricTester):
-    atol = 1e-3  # TODO: ideally tests should pass with lower tolerance
+    atol = 6e-5
 
     # TODO: for some reason this test hangs with ddp=True
     # @pytest.mark.parametrize("ddp", [True, False])
```
Comment (teddykoker) [inline tests/metrics/regression/test_ssim.py:57]: Do tests pass with ddp=True?

File: pytorch_lightning/metrics/functional/ssim.py (hunk 2 above)
Comment (teddykoker) [reply pytorch_lightning/metrics/functional/ssim.py:24]: Also agree with this, otherwise LGTM, thanks!

**Review by teddykoker — CHANGES_REQUESTED** [review]

File: tests/metrics/regression/test_ssim.py (hunk 3 above)
Comment (ydcjeff) [reply tests/metrics/regression/test_ssim.py:57]: The test hangs with ddp=True

Comment (SkafteNicki) [reply tests/metrics/regression/test_ssim.py:57]: @ydcjeff I just tried locally and the test does not hang for me when `ddp=True`.

Comment (ydcjeff) [reply tests/metrics/regression/test_ssim.py:57]: When I tried before, the test hangs on drone, will try again.

Comment (SkafteNicki) [reply tests/metrics/regression/test_ssim.py]: Okay test still hangs on drone, so lets just disable ddp. Really not sure why it is not working.

**Review by teddykoker — APPROVED** [review]
DDP tests hang locally for me, but I guess its fine
